### PR TITLE
Implement and test upsert

### PR DESCRIPTION
**Pull request by elias-ba**
Implemented upsert using pattern `upsert(resourceType, query, data, options = [}, callback = false)` and wrote 3 integration tests for it.

### CODE REVIEW

**Review by taylordowns2000 — CHANGES_REQUESTED** [review]
This looks great @elias-ba . I've got three line-specific questions (see in-line review) that I'd like you to address, and then I'd like a standard test (not an integration test) to be built to lock in this functionality. Apologies here—I know that earlier I said that I didn't think a mocked test would be valuable but not I'm realizing that the behaviour here is wonky enough that we really need one.

I've implemented the first, laying down a pattern [here](https://github.com/OpenFn/language-dhis2/blob/bce9fa3f13233672ecb6c323081baf6ae50b86ee/test/index.js#L250-L299). Can you implement the final two cases?

**Comment by taylordowns2000** [comment]
inline comments resolved in the last 3 commits (per our call). only 2 tests remain

File: src/Adaptor.js [hunk 1]
```diff
@@ -453,55 +453,64 @@ export function get(resourceType, query, options = {}, callback = false) {
...
+  data,
+  options = {},
+  callback = false
+) {
+  return state => {
+    console.log(`Preparing upsert via 'get' then 'create' OR 'update'...`);
+
+    return get(
+      resourceType,
+      query
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply src/Adaptor.js]: Done

File: src/Adaptor.js [hunk 2]
```diff
@@ -453,55 +453,64 @@ export function get(resourceType, query, options = {}, callback = false) {
...
+    )(state).then(resp => {
+      const resources = resp.data[resourceType];
+      if (resources.length > 1) {
+        throw new RangeError(
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply src/Adaptor.js]: Thanks

File: src/Adaptor.js [hunk 3]
```diff
@@ -453,55 +453,64 @@ export function get(resourceType, query, options = {}, callback = false) {
...
+          `Cannot upsert on Non-unique attribute. The operation found more than one records for your request.`
+        );
+      } else if (resources.length <= 0) {
+        return create(resourceType, data, options, callback)(state);
+      } else {
+        const pathName =
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply src/Adaptor.js]: Yes

**Comment by elias-ba** [comment]
@taylordowns2000 all the change requests are now addressed. I also have added the unit tests for upsert using nock. See my last commit

**Pull request merged**